src/Nethermind/Nethermind.Evm/VirtualMachine.cs

```diff
@@ -2393,6 +2396,22 @@ void UpdateMemoryCost(in UInt256 position, in UInt256 length)
                             }
 
                             Span<byte> initCode = vmState.Memory.LoadSpan(in memoryPositionOfInitCode, initCodeLength);
+                            // if container is EOF init code must be EOF
+                            if (spec.IsEip3540Enabled)
+                            {
+                                EofHeader initcodeHeader = null;
+                                bool hasEofPrefix = _byteCodeValidator.HasEOFMagic(initCode);
+                                var isInitcodeEof = hasEofPrefix && _byteCodeValidator.ValidateBytecode(initCode, spec, out initcodeHeader);
+                                bool isValid = CodeContainer.IsEof.Value
+                                    ? isInitcodeEof && CodeContainer.Header?.Version == initcodeHeader?.Version
+                                    : !hasEofPrefix;
```

Comment:
this will disallow non EOF code from deploying EOF code which is not in the spec.